src/blog/understanding-ecmascript-part-3.md

```diff
@@ -0,0 +1,322 @@
+
+> [<code>VariableDeclarationList<sub>[In, Yield, Await]</sub> :</code>](https://tc39.es/ecma262/#prod-VariableDeclarationList)
+> <code>VariableDeclaration<sub>[?In, ?Yield, ?Await]</sub></code>
+
+(Here we show only the production relevant to our example.)
+
+> [<code>VariableDeclaration<sub>[In, Yield, Await]</sub> :</code>](https://tc39.es/ecma262/#prod-VariableDeclaration)
+> <code>BindingIdentifier<sub>[?Yield, ?Await]</sub> Initializer<sub>[?In, ?Yield, ?Await]</sub> opt</code>
+
+The `opt` shorthand means that the right-hand side symbol is optional; there are in fact two productions, one with the optional symbol, and one without.
+
+In the simple case relevant to our example, `VariableStatement` consists of the keyword `var`, followed by a single `BindingIdentifier` without an initializer, and ending with a semicolon.
+
+To disallow or allow `await` as a `BindingIdentifier`, we hope to end up with something like this:
+
+> `BindingIdentifier_Await :`
+> `Identifier`
+> `yield`
+>
+> `BindingIdentifier :`
+> `Identifier`
+> `yield`
+> `await`
+
+This would disallow `await` as an identifier inside async functions and allow it as an identifier inside non-async functions.
+
+But the spec doesn't define it like this, instead we find this production:
+
+> [<code>BindingIdentifier<sub>[Yield, Await]</sub> :</code>](https://tc39.es/ecma262/#prod-BindingIdentifier)
+> `Identifier`
+> `yield`
+> `await`
+
+Expanded, this means the following productions:
+
+> `BindingIdentifier_Await :`
+> `Identifier`
+> `yield`
+> `await`
+>
+> `BindingIdentifier :`
+> `Identifier`
+> `yield`
+> `await`
+
+(We're omitting the productions for `BindingIdentifier_Yield` and `BindingIdentifier_Yield_Await` which are not needed in our example.)
+
+This looks like `await` and `yield` would be always allowed as identifiers. What's up with that? Is the whole blog post for nothing?
+
+### Statics semantics to the rescue
+
+Turns out **static semantics** are needed for forbidding `await` as an identifier inside async functions.
+
+Static semantics describe static rules &mdash; that is, rules that can be checked before the program is ran.
+
+In this case, the [static semantics for BindingIdentifier](https://tc39.es/ecma262/#sec-identifiers-static-semantics-early-errors) define the following syntax-directed rule:
+
+>`BindingIdentifier : await`
+>
+> It's a Syntax Error if this production has an <code><sub>[Await]</sub></code> parameter.
+
+Effectively, this forbids the `BindingIdentifier_Await : await` production.
+
+The spec explains that the reason for having this production but defining it as a Syntax Error by the static semantics is because of interference with automatic semicolon insertion. If the production was missing, automatic semicolon insertion might kick in and insert a semicolon into a program which is syntactically incorrect only because it uses `await` or `yield` as an identifier, changing the meaning of the program.
```

Comment:
Actually, do you think this paragraph above is correct? :) I'm now doubting it. The part "incorrect **only** because it uses...". Is it possible that the *only* error is that await was used as a keyword, and if the production was disallowed, ASI would turn the program into something correct?

The spec uses this as an example:

let
await 0;

and ASI turns it into 

let;
await 0;

But here it's not the case that the only error was "await" as an identifier: let await 0; is also incorrect as a let statement.

However, I don't want to use that example since it has "let" as an identifier too and I don't want to go there. So if what I'm saying is correct, that it's possible to have a program where the *only* error is await as an identifier, and ASI changing the meaning, that would be an easier to understand example.

However, I think it's maybe not possible...

let
await;

is not correct even after ASI so hmm.

(Sorry, totally rubberducking here, I think the answer will be clear to me once I post this comment.)

Reply:
Good call, I didn't think it through that closely.

Here is the original bug: https://github.com/tc39/test262/pull/956#issue-114912013

The problem was that previously, ASI was being made dependent on whether or not `await` was allowed as an identifier. So in async functions, where identifiers can't be named `await`, `let \n await 0;` ended up being parsed instead of raising an Early Error. This was deemed super confusing. 

*With* the static semantics rule, the error is that `await` is used as an identifier. *Without* the static semantics rule, it'll just a `let; await 0;` which is legal syntax. So I think that sentence is inverted. Maybe something like:

"If the production was missing, automatic semicolon insertion will kick in and transform the program into the legal program `let; await 0;` in contexts where `await` is disallowed as an identifier, like inside async functions. The rule aims to prevent this transformation, which TC39 judged to be egregiously confusing."

Reply:
Rewrote this part.